build: Fix guix linker-loader path and add check_ELF_interpreter

dongcarl opened the pull request:
Fixes https://github.com/bitcoin/bitcoin/issues/23111

It would seem that I got the wrong default glibc-dynamic-linker path for PowerPC platforms. This means that for our currently released v22.0 binaries to be run on powerpc platforms, users would have to either:
1. Move `/lib64/ld64.so.?` to `/lib`, or
2. Invoke their linker-loader directly to start our binaries, e.g. `/lib64/ld64.so.? bitcoind`

This is my bad.

I've fixed the paths in this patchset, and also added a test to `symbol-check.py` so that this does not ever slip past our checks again.

dongcarl committed:
guix: Fix powerpc64(le) dynamic linker name

I used Guix's values for the powerpc64(le) dynamic linkers, and the
/lib-prefix seems to be a Guix-ism rather than standard. The standard
path for the linker-loaders start with /lib64.

I've taken the new loader values from SYSDEP_KNOWN_INTERPRETER_NAMES in
glibc's sysdeps/unix/sysv/linux/powerpc/ldconfig.h file.

For future reference, loader path values can also be found on glibc's
website: https://sourceware.org/glibc/wiki/ABIList?action=recall&rev=16

dongcarl committed:
symbol-check: Check requested ELF interpreter

It is important that binaries request a standard interpreter location
where most distros would place the linker-loader. Otherwise, the user
would be met with a very confusing message:

    bash: <path>/<to>/bitcoind: No such file or directory

When really it's the interpreter that's not found.

laanwj commented on contrib/devtools/symbol-check.py:
```diff
@@ -244,11 +259,23 @@ def check_PE_subsystem_version(filename) -> bool:
 
+def check_ELF_interpreter(filename) -> bool:
+    binary = lief.parse(filename)
+    elf = pixie.load(filename)
```
using both lief and pixie seems overkill here, please choose one :smile:

fanquake replied:
I would opt for `lief` as we are moving that direction #22392.

dongcarl replied:
I explained this in the commit message:

```
Code reviewers: You may see that both lief and pixie are used for
check_ELF_interpreter. That's because at the time of writing:
  - lief does not recognize riscv64 yet (fixed by fanquake in e83f812,
    not yet in a release)
  - pixie does not know how to extract the interpreter
```

Let me know if I'm missing some functionality here!

laanwj replied:
Oops, I missed that commit message, sorry.

laanwj commented:
Concept ACK, this is a good check to have.

dongcarl commented:
Pushed e95b5a9ca9040cab826f041ab7cf348696a15ba3 -> 2261b186ebae38362eaceeed5ac1991b578375fa
- Cherry-picked 2261b186ebae38362eaceeed5ac1991b578375fa from #22392 as e759330d003099a8acacb1e2b7da00e1f3f1d1ea
    - Modifications:
        - Use `lief.ELF.ARCH(243)` instead of 'RISCV' special value
        - Update `assertEqual` expected string in `test-symbol-check`
- Stop using `pixie` for the new ELF interpreter check

laanwj commented:
I missed this before, but I think it can now be removed from the commit message of 2261b186ebae38362eaceeed5ac1991b578375fa, as only lief is used now?
```
    Code reviewers: You may see that both lief and pixie are used for
    check_ELF_interpreter. That's because at the time of writing:
      - lief does not recognize riscv64 yet (fixed by fanquake in e83f812,
        not yet in a release)
      - pixie does not know how to extract the interpreter
```

dongcarl commented:
> can now be removed from the commit message

Yes, done in latest push!

laanwj commented:
Thank you
Code review ACK 6071282e0af299ebc90ebc8d1b4662a081574bf5

dongcarl commented:
Pushed b74a251e042b8d09c34fd452613736edc339ed30 -> 1527b7e8a1339705a48d4d1ed108230fc4953c18
- Rebased on master after merge of #22392
- Addressed https://github.com/bitcoin/bitcoin/pull/23148#discussion_r728016267
